cgen: reusing first variant

### Describe the bug

<title>

### Reproduction Steps

```v
struct Context {
}

struct Input {
}

struct Block {
	x int
	y int
mut:
	children []Element
}

type Element = Block | Input

fn (input Input) draw(mut ctx Context, x int, y int) {}

fn (block Block) draw(mut ctx Context, x int, y int) {
	for child in block.children {
		child.draw(mut ctx, block.x + x, block.y + y)
	}
}

fn (elt Element) draw(mut ctx Context, x int, y int) {
	$for T in Element.variants {
		if elt is T {
			elt.draw(mut ctx, x, y)
			return
		}
	}
}

fn main() {}
```

### Expected Behavior

Successful compile

### Current Behavior

```
==================
C:/Users/mclr/AppData/Local/Temp/v_0/test.01HP0FW0CW2200DDM2E6M2Q72H.tmp.c:6903: warning: implicit declaration of function 'tcc_backtrace'
C:/Users/mclr/AppData/Local/Temp/v_0/test.01HP0FW0CW2200DDM2E6M2Q72H.tmp.c:13115: error: cannot convert 'struct main__Block' to 'struct main__Input'
...
==================
(Use `v -cg` to print the entire error message)

builder error:
==================
C error. This should never happen.

This is a compiler bug, please report it using `v bug file.v`.

https://github.com/vlang/v/issues/new/choose

You can also use #help on Discord: https://discord.gg/vlang

```

### Possible Solution

_No response_

### Additional Information/Context

_No response_

### V version

V 0.4.4 9f32af7

### Environment details (OS name and version, etc.)

```
V full version: V 0.4.4 57a33c3.9f32af7
OS: windows, Microsoft Windows 11 Pro v22000 64-bit
Processor: 12 cpus, 64bit, little endian, 

getwd: D:\Games\Proekti\V\interactions
vexe: D:\Games\Proekti\V\v\v.exe
vexe mtime: 2024-02-07 00:34:15

vroot: OK, value: D:\Games\Proekti\V\v
VMODULES: OK, value: C:\Users\mclr\.vmodules
VTMP: OK, value: C:\Users\mclr\AppData\Local\Temp\v_0

Git version: git version 2.37.0.windows.1
Git vroot status: weekly.2024.05-36-g9f32af78
.git/config present: true

CC version:
thirdparty/tcc status: thirdparty-windows-amd64 a39eb79b
```

--- a/vlib/v/gen/c/fn.v
+++ b/vlib/v/gen/c/fn.v
@@ -1307,10 +1307,14 @@ fn (mut g Gen) resolve_receiver_type(node ast.CallExpr) (ast.Type, &ast.TypeSymb
 	if node.from_embed_types.len == 0 && node.left is ast.Ident {
 		if node.left.obj is ast.Var {
 			if node.left.obj.smartcasts.len > 0 {
-				unwrapped_rec_type = g.unwrap_generic(node.left.obj.smartcasts.last())
-				cast_sym := g.table.sym(unwrapped_rec_type)
-				if cast_sym.info is ast.Aggregate {
-					unwrapped_rec_type = cast_sym.info.types[g.aggregate_type_idx]
+				if node.left.obj.ct_type_var == .smartcast {
+					unwrapped_rec_type = g.unwrap_generic(g.comptime.get_comptime_var_type(node.left))
+				} else {
+					unwrapped_rec_type = g.unwrap_generic(node.left.obj.smartcasts.last())
+					cast_sym := g.table.sym(unwrapped_rec_type)
+					if cast_sym.info is ast.Aggregate {
+						unwrapped_rec_type = cast_sym.info.types[g.aggregate_type_idx]
+					}
 				}
 			}
 		}

--- a/vlib/v/tests/comptime_smart_receiver_test.v
+++ b/vlib/v/tests/comptime_smart_receiver_test.v
@@ -1,0 +1,35 @@
+struct Context {
+}
+
+struct Input {
+}
+
+struct Block {
+	x int
+	y int
+mut:
+	children []Element
+}
+
+type Element = Block | Input
+
+fn (input Input) draw(mut ctx Context, x int, y int) {}
+
+fn (block Block) draw(mut ctx Context, x int, y int) {
+	for child in block.children {
+		child.draw(mut ctx, block.x + x, block.y + y)
+	}
+}
+
+fn (elt Element) draw(mut ctx Context, x int, y int) {
+	$for T in Element.variants {
+		if elt is T {
+			elt.draw(mut ctx, x, y)
+			return
+		}
+	}
+}
+
+fn test_main() {
+	assert true
+}

